ingest/ingest_retry_client.go

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2023 IndyKite
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ingest
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"google.golang.org/grpc/codes"
+
+	"github.com/indykite/indykite-sdk-go/errors"
+	ingestpb "github.com/indykite/indykite-sdk-go/gen/indykite/ingest/v1beta2"
+	api "github.com/indykite/indykite-sdk-go/grpc"
+)
+
+type RetryPolicy struct {
+	InitialBackoff    string
+	CtxTimeout        time.Duration
+	MaxAttempts       int
+	BackoffMultiplier int
+}
+
+type RetryClient struct {
+	*Client
+	retryPolicy         *RetryPolicy
+	isUnableToReconnect bool
+}
+
+// NewRetryClient creates a new Ingest API gRPC Client with retry functionality.
+func NewRetryClient(ctx context.Context, retryPolicy *RetryPolicy, opts ...api.ClientOption) (*RetryClient, error) {
+	client, err := NewClient(ctx, opts...)
+	if err != nil {
+		return nil, err
+	}
+	c := &RetryClient{
+		Client:      client,
+		retryPolicy: retryPolicy,
+	}
+
+	if _, err = time.ParseDuration(c.retryPolicy.InitialBackoff); err != nil {
+		return nil, errors.New(codes.FailedPrecondition, fmt.Sprintf("invalid backoff duration: %v", err))
+	}
+	return c, nil
+}
+
+// NewTestRetryClient creates a new Ingest API gRPC Client with retry functionality.
+func NewTestRetryClient(client ingestpb.IngestAPIClient, retryPolicy *RetryPolicy) (*RetryClient, error) {
+	c, err := NewTestClient(client)
+	if err != nil {
+		return nil, err
+	}
+	if err != nil {
+		return nil, err
+	}
+	rc := &RetryClient{
+		Client:      c,
+		retryPolicy: retryPolicy,
+	}
+
+	if _, err = time.ParseDuration(rc.retryPolicy.InitialBackoff); err != nil {
+		return nil, errors.New(codes.FailedPrecondition, fmt.Sprintf("invalid backoff duration: %v", err))
+	}
+	return rc, nil
+}
+
+// SendRecord sends a record on the opened stream and retries if it fails.
+func (c *RetryClient) SendRecord(record *ingestpb.Record) error {
+	if c.stream == nil {
+		return errors.New(codes.FailedPrecondition, "a stream must be opened first")
+	}
+
+	err := c.Client.SendRecord(record)
+	if err == nil {
+		return err
```

Comment:
Not sure what the best practice is here, but at first glance it's a little confusing to read `return err` when the `err` is in fact `nil` in this case. Would it be better to `return nil` here?

Reply:
The same applies to a few other occurrences.

Reply:
fixed